TypeSystem/flow-analysis/promotion_via_assignment_A05_t01.dart

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion We say that a variable `x` is promotable via assignment of an
+/// expression of type `T` given variable model `VM` if
+/// - `VM = VariableModel(declared, promoted, tested, assigned, unassigned,
+///     captured)`
+/// - and captured is false
+/// - and `S` is the current type of `x` in `VM`
+/// - and `T <: S` and not `S <: T`
+/// - and `T` is a type of interest for `x` in `tested`
+///
+/// @description Checks that if a type of variable is not specified and there is
+/// no initializer then its type is `dynamic` in spite of assignment.
```

Comment:
I don't think there is anything special about `dynamic` here. We used to have a rule saying that a variable of type `dynamic` is never promoted (the idea is that this would _take away_ member invocations that you'd be allowed to perform, and promotion should never eliminate any allowed constructs), but we stopped having that rule.

This matches the behavior that I can see, too: A variable of type `dynamic` can be promoted by a type test, just like a variable of any other type (`dynamic d = 1; if (d is int) ...`).

Reply:
Thanks!